cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -1,55 +1,247 @@
 };
-use std::collections::HashSet;
+use std::{collections::HashSet, pin::Pin};
 
 type Hash = H256;
 type Block = crate::ethereum::Block<Transaction>;
 
-pub async fn matching_transaction<C>(
+pub const TRANSACTION_STATUS_OK: u32 = 1;
+
+pub async fn matching_create_contract_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    bytecode: Bytes,
+) -> anyhow::Result<(Transaction, Address)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    let (transaction, receipt) =
+        matching_transaction(blockchain_connector, start_of_swap, |transaction| {
+            // transaction.to address is None if, and only if, the transaction
+            // creates a contract.
+            transaction.to.is_none() && transaction.input == bytecode
+        })
+        .await?;
+
+    match receipt.contract_address {
+        Some(location) => Ok((transaction, location)),
+        None => return Err(anyhow::anyhow!("contract address missing from receipt")),
+    }
+}
+
+pub async fn matching_event_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    event: Event,
+    action: &'static str,
+) -> anyhow::Result<(Transaction, TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction_bloom_filter(
+        blockchain_connector.clone(),
+        start_of_swap,
+        event.topics.clone(),
+        action,
+        |transaction| {
+            Box::pin({
+                let blockchain_connector = blockchain_connector.clone();
+                let event = event.clone();
+                async move {
+                    let receipt = fetch_receipt(blockchain_connector, transaction.hash).await?;
+                    if event_exists_in_receipt(&event, &receipt) {
+                        return Ok(Some((transaction, receipt)));
+                    }
+
+                    Ok(None)
+                }
+            })
+        },
+    )
+    .await
+}
+
+/// Fetch receipt from connector using transaction hash.
+async fn fetch_receipt<C>(blockchain_connector: C, hash: Hash) -> anyhow::Result<TransactionReceipt>
+where
+    C: ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>,
+{
+    let receipt = blockchain_connector
+        .receipt_by_hash(hash)
+        .compat()
+        .await?
+        .ok_or_else(|| {
+            anyhow::anyhow!(
+                "Could not get transaction receipt for transaction {:x}",
+                hash
+            )
+        })?;
+    Ok(receipt)
+}
+
+fn event_exists_in_receipt(event: &Event, receipt: &TransactionReceipt) -> bool {
+    match event {
```

Comment:
I think it would be the right time to refactor this method to use defensive programming:

```rust
let { topics, address } = event;

if topics.is_empty() {
   return false;
}

// etc ...
```

Reply:
Oh, you think that is nicer?  Interesting, that's how I had it for a lot time I was working on this.  I refactored it to use the 'if guard' within the match.  Unless I'm missing the meaning of your comment?
```
    match event {
        Event { topics, .. } if topics.is_empty() => false,
```
Is the same as :
```
let { topics, address } = event;

if topics.is_empty() {
   return false;
}
```

Reply:
Yes it is the same functionally :)
But the last time I did that, you were like: "that is just using the language features for the sake of it! :D"

And then you suggested a defensive programming approach which is also quite neat.
I am not sure what is better in this case honestly. The defensive one maps clearer to the algorithm that one would have in their mind I guess :man_shrugging: 

Up to you - no strong feelings on this part :)